### ISSUE TITLE
Question(s): Many playbooks , many adhoc playbooks, multiple roles / monolith repository, how|where does molecule fit?

### ISSUE THREAD

**Issue by claudio-viola** [metacloud/molecule#1010]
First of all thanks for your great work.

# Issue Type

- Question
This is more of a general question to understand the best practice that molecule tries to enforce.

I think it would be great to have more support for so called "monolith repositories" 
(I have read this one https://molecule.readthedocs.io/en/latest/examples.html#monolith-repo)
I for instance would prefer the approach of  keeping the test(s) code separate from the roles and in a monolith repository it's not really possible with molecule or is it?

root/ 
 playbook/
  play1.yml
  play2.yml
  play3.yml
 roles/
   role1
   role2
   role3
 test/
 scenario1/
   role1/
   role2/
scenario2/
   role1/
   role2/
   role3/

 but i understand that molecule enforces testing roles in isolation so for a monolith repo the following structure should be used
roles/
 role1/
 molecule/
  scenario1/
     ./molecule.yml 
  scenario2/
   ./molecule/
     ./molecule.yml 
  role2/
  ./molecule/
    ./molecule.yml 
  role3/
    ./molecule/
    ./molecule.yml 

It is mentioned that molecule is inspired by "kitchen" so **just to understand** is there a fundamental reason the approach in molecule is of "testing roles in isolation" ?
 
Testing "monolith" repository with multiple roles 
inventories/
 inventory1
 inventory2
roles/
role1
role2
role3

Would the best practice here to "unbundle" the monolith repository into separate independent roles? What happens to adhoc playbooks?

# Desired Behaviour
In root of an ansible "monolith" repository  be able to run
molecule test  
to runs all tests/scenarios against all roles of a monolith repository

Basically like kitchen would be great to be able to run  suites / scenarios  as a top level feature that encompasses includes roles.

Inspired by the kitchen concept of suites
suites:
  - name: default
    run_list:
      - recipe[recipe1::default]  ~~~ role[1]
      - recipe[recipe2::default]  ~~~ role[2]

Not sure If i managed to give the idea but basically what's the best practice for testing monolith repository with multiple roles , multiple playbooks in general ? 
Is there a plan for better managing this kind of repositories?

**Comment by retr0h** [comment]
You can have a monolith repo, but you still need molecule initialized inside each role you are planning to test and you still need to change directory into the role you are planning to test to execute Molecule.  Yes, in general molecule likes to test roles in isolation, but you are still free to compose roles which rely on other roles.

We actually have a monolith repo at work.  Molecule lives in each role, and we change into the directory before we bug fix or feature add.  However, we do run into issues with CI in this layout b/c we only want to test the roles which changed and/or roles which depend on the role changed.   You can share the bootstrap playbooks across roles in your monolith, so you don't need to redefine the create/destroy files [1].

To answer your questions scenarios are not a top level feature.  Scenarios are treated on a per role level.  If a role is to test other roles for a particular scenario, then I create a scenario inside a deployment "meta" role. 

[1] https://github.com/retr0h/ansible-etcd/blob/master/molecule/ubuntu/molecule.yml#L29-L32

**Comment by claudio-viola** [comment]
Thanks for the answer @retr0h  I think I understand. Exactly those are the issues  I was foreseeing a nd trying to workaround them. CI and having multiple molecule folders per role and  run  a "molecule test" for each role.
Is there any valid work around to run molecule run tests against a role from not within the role folder ?

**Comment by retr0h** [comment]
> Is there any valid work around to run molecule tests against a role from not within the role folder ?

No, it is not designed to function that way.

**Comment by claudio-viola** [comment]
thank you so much , great job. can be closed for me! 👍
